### PR TITLE
[feature] Hill cipher algorithm

### PR DESCRIPTION
**Pull request by kvedala**
#### Description of Change

* New folder for cipher algorithm implementations
* Implementation of Hill Cipher algorithm

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

File: ciphers/hill_cipher.cpp [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+template <typename T>
+static std::ostream &operator<<(std::ostream &out, matrix<T> const &v) {
+    const int width = 15;
+    const char separator = ' ';
+
+    for (size_t row = 0; row < v.size(); row++) {
+        for (size_t col = 0; col < v[row].size(); col++)
+            out << std::left << std::setw(width) << std::setfill(separator)
+                << v[row][col];
+        out << std::endl;
```
Comment (Panquesito7) [inline ciphers/hill_cipher.cpp:37]:
```suggestion
        for (size_t col = 0; col < v[row].size(); col++) {
            out << std::left << std::setw(width) << std::setfill(separator)
                << v[row][col];
        }
        out << std::endl;
```
Same in other parts of the code.

Comment (kvedala) [reply ciphers/hill_cipher.cpp:37]: do not worry about it for now. This is a work in progress - a draft pull.

Comment (kvedala) [reply ciphers/hill_cipher.cpp:37]: @Panquesito7 wrote 400+ lines of code without a single LGTM alert :)

**Comment by kvedala** [comment]
sample output as of fc489f7
```
Key dictionary: (97)
        ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789~!@#$%^&*()_+`-=[]{}|;':",./<>?\

======Test 1 (3x3 key) ======
Original text:
        This is a simple text with numb3r5 and exclamat!0n.
Encrypted text:
u       'a*'o2`>Hu3)q`S!ZJb7Kv}KU+Q']7#^Q\6EwL5Ui?Y$(Jf>L
Reconstruct text:
        This is a simple text with numb3r5 and exclamat!0n.
Passed :)
======Test 2 (8x8 key) ======
Original text:
        This is a simple text with numb3r5 and exclamat!0n.
Encrypted text:
        b>TK'T3 }a#PEj(nn wBTW|n(s6#K<BxN`;woLX{9K ;Iu:=ZxTZDrb8
Reconstruct text:
        This is a simple text with numb3r5 and exclamat!0n.AAAAA
Passed :)
```

**Comment by kvedala** [comment]
> This pull request **introduces 3 alerts** when merging [25c2b82](https://github.com/TheAlgorithms/C-Plus-Plus/commit/25c2b82e51b3311998973168e2a9b9c0a3f643c7) into [eddda4e](https://github.com/TheAlgorithms/C-Plus-Plus/commit/eddda4e43ebe1ba797a5b31d962515bd56d3888e) - [view on LGTM.com](https://lgtm.com/projects/g/TheAlgorithms/C-Plus-Plus/rev/pr-1e91d48def8d679bf2d29912f6b27205c6d2284d)
> 
> **new alerts:**
> 
> * 2 for Constant return type on member
> * 1 for Comparison of narrow type with wide type in loop condition

Fixed

**Pull request merged**